plugins/util/html_emitter.py

```diff
@@ -0,0 +1,365 @@
+            {"class":"module-type-text"},2,True)
+          idents_div += ht.para(ht.add_tag("span","description:",{"class":"module-type-text-label"}) + ht.br(newline=True) + id.attrs['desc'],{"class":"module-type-text"},2,True)
+
+        idents_div += ht.close_tag(newline=True)
+    else:
+      # module doesn't have any identities
+      idents_div = ""
+
+    # store the identity docs
+    self.moduledocs[mod.module_name]['identities'] = idents_div
+
+    gen_nav_tree(self, mod, 0)
+
+
+  def genStatementDoc(self, statement, ctx, level=1):
+    """HTML emitter for module data node given a StatementDoc
+    object"""
+
+    if ctx.opts.no_structure and statement.keyword in ctx.skip_keywords:
+      return
+
+    ht = html_helper.HTMLHelper()
+
+    s_div = ht.open_tag("div", {"class":"statement-section"}, newline=True)
+
+    if ctx.opts.strip_namespace:
+      pathstr = yangpath.strip_namespace(statement.attrs['path'])
+    else:
+      pathstr = statement.attrs['path']
+
+    # for 'skipped' nodes, just print the path
+    if statement.keyword in self.path_only:
+      s_div += ht.h4(pathstr,None,level,True)
+      s_div += ht.close_tag(newline=True)
+      return s_div
+
+    # statement name and nodetype
+    s_div += ht.h4(statement.name, {"class":"statement-name","id":node_to_id(statement)},level,True)
+
+    # node description
+    if statement.attrs.has_key('desc'):
+      s_div += ht.para(ht.add_tag("span", "description",{"class":"statement-info-label"}) + ":<br />" + statement.attrs['desc'],{"class":"statement-info-text"},level,True)
+    s_div += ht.close_tag(newline=True)
+
+    # check for additional properties
+    notes = ""
+    if statement.attrs['is_key']:
+      notes += " (list key)"
+    if statement.attrs['config']:
+      notes += " (rw)"
+    else:
+      notes += " (ro)"
+    keyword = statement.keyword + notes
+    s_div += ht.para(ht.add_tag("span", "nodetype",{"class":"statement-info-label"}) + ": " + keyword,{"class":"statement-info-text"},level,True)
+    # s_div += ht.para(ht.add_tag("span", "path",{"class":"statement-info-label"}) + ": " + pathstr,{"class":"statement-info-text"},level,True)
+
+    if statement.typedoc:
+      s_div += gen_type_info(statement.typedoc, level)
+
+    for prop in YangDocDefs.type_leaf_properties:
+      if statement.attrs.has_key(prop):
+        s_div += ht.para(ht.add_tag("span", prop, {"class":"statement-info-label"}) + ": " + statement.attrs[prop],{"class":"statement-info-text"},level,True)
+
+    # add this statement to the collection of data
+    self.moduledocs[statement.module_doc.module_name]['data'] += s_div
+
+  def emitDocs(self, ctx, section=None):
+    """Return the HTML output for all modules,
+    or single section if specified"""
+
+    ht = html_helper.HTMLHelper()
+
+    docs = []
+    navs = []
+    navids = []
+    # create the documentation elements for each module
+    for module_name in self.moduledocs.keys():
```

Comment:
FYI, you should just do `for module_name in self.moduledocs` in general. A dict has an iterator that has just the keys by default, if you use `.keys()` for this then it creates a list object of those keys, which generally performs worse.

Since there's lots of dict access here too, it can be cleaner to do `for module_name, module_definition in self.moduledocs.iteritems()` which gives you both the name and then a `module_definition` object, which is the value of the dict entry.